netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java

```diff
@@ -334,16 +335,18 @@ protected ClientTransportFactory buildTransportFactory() {
     TransportCreationParamsFilterFactory transportCreationParamsFilterFactory =
         dynamicParamsFactory;
     if (transportCreationParamsFilterFactory == null) {
-      SslContext localSslContext = sslContext;
-      if (negotiationType == NegotiationType.TLS && localSslContext == null) {
-        try {
-          localSslContext = GrpcSslContexts.forClient().build();
-        } catch (SSLException ex) {
-          throw new RuntimeException(ex);
+      ProtocolNegotiator negotiator = this.protocolNegotiator;
+      if (negotiator == null) {
+        SslContext localSslContext = sslContext;
+        if (negotiationType == NegotiationType.TLS && localSslContext == null) {
+          try {
+            localSslContext = GrpcSslContexts.forClient().build();
+          } catch (SSLException ex) {
+            throw new RuntimeException(ex);
+          }
         }
+        negotiator = createProtocolNegotiatorByType(negotiationType, localSslContext);
```

Comment:
I would feel a lot more comfortable with this if the lifetime of protocol negotiators was documented.  Up until now, it's been ambiguous if there is one per channel, channel builder, transport, etc.   I'm not so keen on which one it is, as long as it's easily understandable.

I'm uncomfortable with creating a new PN each time, except if it has been previously set by outside.  I would have expected `this.protocolNegotiator` to be set here, and cleared out by changing the SSL settings.  The reason is that it's unclear what will happen if SSL related settings are changed on the builder, *after* `protocolNegotiator()` is called.